aten/src/ATen/native/DistributionTemplates.h

```diff
@@ -322,6 +322,9 @@ Tensor& exponential_impl_(Tensor& self, double lambda, c10::optional<Generator>
 
 template<template<typename> class cauchy_kernel, typename RNG>
 Tensor& cauchy_impl_(Tensor& self, double median, double sigma, c10::optional<Generator> gen) {
+  // TODO: instead of variable name 'sigma', use 'gamma' or 'scale'
+  // the variance, squared sigma, is undefined for cauchy distribution
+  TORCH_CHECK(sigma > 0.0, "cauchy_ expects sigma > 0.0, but found sigma=", sigma);
```

Comment:
Why restrict to `sigma > 0` rather than `sigma >= 0`?  I would expect `.cauchy_(median, sigma=0)` to behave as `.fill_(median)`.

Reply:
Thanks @fritzo, I have two comments regarding `sigma = 0`. 

(1) `lim_{sigma -> 0} F^{-1} (x ; sigma, median) = median`. This is perfectly sound for as `sigma -> 0`.
(2) both the PDF and CDF of cauchy distribution are not sound when `sigma = 0`

Considering the two above, sure we _can_ define `sigma = 0` as the behavior at limit as `sigma -> 0`, but is not the same as `sigma = 0`, and PDF, CDF are collapsed for `sigma = 0`.

Reply:
OTOH, this is a sampler, and the distribution of samples converges.  I'd imagine this function would be used for initialization, and it's sure nice if I can set `sigma=0` rather than add `if ... else` logic to decide whether to inject noise.

(I think this less important, but the CDF does converge pointwise almost everywhere to the step function)

Reply:
> this is a sampler, and the distribution of samples converges


Cauchy distribution doesn't obey the law of large numbers, the sample mean doesn't converge to a finite number reflecting that the expected value is undefined for cauchy, instead converges trivially to a cauchy random variable. I don't think it's sound to return `.fill_(median), for sigma =0` and `mean({x_1, x_2, ..., x_n}) = median`. 

> I'd imagine this function would be used for initialization, and it's sure nice if I can set sigma=0 rather than add if ... else logic to decide whether to inject noise.

I'm not sure which applications popularly use cauchy random variable but I would guess many don't soundly set `sigma = 0`. Intuitively, a distribution of cutted distances not cut ?

More importantly, the derivation of cauchy CDF from `tan θ = x / b` isn't sound for `b = 0`. What would we return for the cdf ? 

So I don't think it's much sound intuitively nor physically to define for `sigma = 0`. What do you think ? 

[reference] https://mathworld.wolfram.com/CauchyDistribution.html

Reply:
I guess my main motivation is to continue satisfying an equation like
```
x.cauchy_(median, sigma) == x.cauchy_().mul_(sigma).add_(median)  # in distribution
```
for any `median` and `sigma > 0`, as the current code satisfies.  I'm worried if we add a check for `sigma > 0`, then we'll bloat user code like this:
```diff
- x.cauchy_(median, sigma)
+ if sigma > 0:
+     x.cauchy_(median, sigma)
+ else:
+     x.zero_()
```
but maybe that's ok 🤔 

I guess in the bigger picture, this is a sampler and not a `Distribution` object, so I'd think it's less important to talk about the CDF or central limit theorem.  I do agree with you that `torch.distributions.Cauchy` should restrict `scale > 0` since that `Distribution` object should support methods like `.log_prob()` that are undefined at `scale == 0`.  Maybe as a policy we could require stricter checks to `torch.distributions` wrappers and use looser checks in `.{distribution}_()` methods?

Reply:
`but maybe that's ok 🤔`

Thanks @fritzo, I think this is OK : ) . At the end of the day, whether `torch.Tensor.cauchy_` is a sampler or `torch.distributions.Cauchy` is a distribution, the generated random variable should follow the cauchy distribution. 

If `X` follows standard cauchy distribution, and `median` a real number and `sigma > 0`, then `Y = median + sigma * X` follows the cauchy distribution. And we won't guarantee any behavior for parameterization outside. What do you think?

Reply:
Sounds good to me.  Thanks for caring about design choices 😄

Reply:
haha, thanks @fritzo, enjoyed our discussion debate : )